pkg/transport/process/pipeline.go

```diff
@@ -0,0 +1,83 @@
+package process
+
+import (
+	"context"
+	"os"
+
+	"archive/tar"
+	"fmt"
+	"io/ioutil"
+
+	cdv2 "github.com/gardener/component-spec/bindings-go/apis/v2"
+)
+
+type resourceProcessingPipelineImpl struct {
+	processors []ResourceStreamProcessor
+}
+
+func (p *resourceProcessingPipelineImpl) Process(ctx context.Context, cd cdv2.ComponentDescriptor, res cdv2.Resource) (*cdv2.ComponentDescriptor, cdv2.Resource, error) {
+	infile, err := ioutil.TempFile("", "out")
+	if err != nil {
+		return nil, cdv2.Resource{}, fmt.Errorf("unable to create temporary infile: %w", err)
+	}
+
+	err = WriteTARArchive(ctx, cd, res, nil, tar.NewWriter(infile))
+	if err != nil {
+		return nil, cdv2.Resource{}, fmt.Errorf("unable to write: %w", err)
+	}
+
+	for _, proc := range p.processors {
+		outfile, err := p.process(ctx, infile, proc)
+		if err != nil {
+			return nil, cdv2.Resource{}, err
+		}
+
+		infile = outfile
+	}
+	defer infile.Close()
+
+	_, err = infile.Seek(0, 0)
+	if err != nil {
+		return nil, cdv2.Resource{}, err
+	}
+
+	processedCD, processedRes, blobreader, err := ReadTARArchive(tar.NewReader(infile))
+	if err != nil {
+		return nil, cdv2.Resource{}, fmt.Errorf("unable to read output data: %w", err)
+	}
+	defer blobreader.Close()
+
+	return processedCD, processedRes, nil
+}
+
+func (p *resourceProcessingPipelineImpl) process(ctx context.Context, infile *os.File, proc ResourceStreamProcessor) (*os.File, error) {
+	defer infile.Close()
+
+	_, err := infile.Seek(0, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to seek to beginning of input file: %w", err)
+	}
+
+	outfile, err := ioutil.TempFile("", "out")
+	if err != nil {
+		return nil, fmt.Errorf("unable to create temporary outfile: %w", err)
+	}
+
+	inreader := infile
+	outwriter := outfile
+
+	err = proc.Process(ctx, inreader, outwriter)
+	if err != nil {
+		return nil, fmt.Errorf("unable to process resource: %w", err)
+	}
+
+	return outfile, nil
```

Comment:
we should also cleanup the in file? at after everything has been finished

Reply:
for every processor in the chain (except the last one), the file should be cleaned up in line 56. I don't clean up the file for the last processor run because it is returned to the calling instance (which then has to perform the cleanup). 
It would also be possible to pass in a writer instance for the blob (so the caller can decide where to write the blob and has to explicitly handle the underlying resources), but in my opinion this would make the function interface look weird, where parts of the result is returned via return values and the other part is written via a writer parameter.